REGRESSION(266828@main): Nested setTimeout not being throttled in the background

rniwa opened the pull request:
#### e49305511b1aee7144154ad1530454b0970ac515
<pre>
REGRESSION(266828@main): Nested setTimeout not being throttled in the background
<a href="https://bugs.webkit.org/show_bug.cgi?id=262086">https://bugs.webkit.org/show_bug.cgi?id=262086</a>
&lt;rdar://115902836&gt;

Reviewed by Chris Dumez.

The bug was caused by TimerAlignment and hasReachedMaxNestingLevel state getting set after
the timer had already been scheduled. We need to update these states before we schedule it.

To do this, new variants of scheduleTask and scheduleRepeatingTask which takes TimerAlignment
and hasReachedMaxNestingLevel boolean as arguments are added and deployed them in DOMTimer.

* LayoutTests/fast/dom/timer-throttling-hidden-page-2-expected.txt: Added.
* LayoutTests/fast/dom/timer-throttling-hidden-page-2.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask):
(WebCore::EventLoop::scheduleRepeatingTask):
(WebCore::EventLoopTaskGroup::scheduleTask):
(WebCore::EventLoopTaskGroup::scheduleRepeatingTask):
(WebCore::EventLoopTaskGroup::setTimerAlignment): Deleted.
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::DOMTimer):

Canonical link: <a href="https://commits.webkit.org/268460@main">https://commits.webkit.org/268460@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d01ec0982285434cafc335f51f6a266f15c87004

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19776 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20199 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21671 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18482 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20011 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23459 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20343 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17200 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22526 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17164 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17995 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24284 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18239 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18169 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22266 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15898 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17925 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17831 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22277 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2414 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18590 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
LGTM though I kinda wish we had an enum class for HasReachedMaxNestingLevel.

rniwa committed:
REGRESSION(266828@main): Nested setTimeout not being throttled in the background
https://bugs.webkit.org/show_bug.cgi?id=262086
<rdar://115902836>

Reviewed by Chris Dumez.

The bug was caused by TimerAlignment and hasReachedMaxNestingLevel state getting set after
the timer had already been scheduled. We need to update these states before we schedule it.

To do this, new variants of scheduleTask and scheduleRepeatingTask which takes TimerAlignment
and hasReachedMaxNestingLevel boolean as arguments are added and deployed them in DOMTimer.

* LayoutTests/fast/dom/timer-throttling-hidden-page-2-expected.txt: Added.
* LayoutTests/fast/dom/timer-throttling-hidden-page-2.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask):
(WebCore::EventLoop::scheduleRepeatingTask):
(WebCore::EventLoopTaskGroup::scheduleTask):
(WebCore::EventLoopTaskGroup::scheduleRepeatingTask):
(WebCore::EventLoopTaskGroup::setTimerAlignment): Deleted.
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::DOMTimer):

Canonical link: https://commits.webkit.org/268460@main
